Sources/StreamChatUI/VoiceRecording/VoiceRecordingVC.swift

```diff
@@ -0,0 +1,684 @@
+            hideViewsDebouncer.invalidate()
+            Animate { [weak self] in
+                guard let self = self else { return }
+
+                self.lockIndicatorView.alpha = 0
+
+                [
+                    self.discardRecordingButton,
+                    self.stopRecordingButtonContainer,
+                    self.recordingIndicatorView,
+                    self.slideToCancelView
+                ].forEach {
+                    if $0.superview == self.centerContainer {
+                        self.centerContainer.removeArrangedSubview($0)
+                    }
+                }
+
+                [self.sendButton, self.confirmButton].forEach {
+                    if $0.superview == self.trailingContainer {
+                        self.trailingContainer.removeArrangedSubview($0)
+                    }
+                }
+
+                self.liveRecordingView.removeFromSuperview()
+
+                self.recordingTipView.alpha = 0
+            } completion: { [weak self] completed in
+                guard completed, let self = self else { return }
+                self.lockIndicatorView.removeFromSuperview()
+                self.lockIndicatorView.bottomPaddingConstraint.constant = self.lockIndicatorView.minimumBottomPadding
+                self.lockIndicatorView.alpha = 1
+                self.recordingTipView.removeFromSuperview()
+            }
+
+        case .showingTip:
+            recordingTipView.alpha = 1
+            Animate { [weak self] in
+                guard let self = self else { return }
+                self.delegate?.voiceRecording(self, presentFloatingView: self.recordingTipView)
+            } completion: { [weak self] completed in
+                guard completed else { return }
+                self?.hideViewsDebouncer.execute { [weak self] in
+                    Animate { [weak self] in
+                        self?.recordingTipView.alpha = 0
+                    } completion: { [weak self] completed in
+                        guard completed else { return }
+                        self?.recordingTipView.removeFromSuperview()
+                    }
+                }
+            }
+
+        case .recording:
+            delegate?.voiceRecordingDidBeginRecording(self)
+
+            lockIndicatorView.isHidden = true
+            lockIndicatorView.alpha = 1
+            delegate?.voiceRecording(self, presentFloatingView: lockIndicatorView)
+            lockIndicatorView.content = false
+            lockIndicatorView.setNeedsLayout()
+
+            Animate { [weak self] in
+                guard let self = self else { return }
+                self.lockIndicatorView.isHidden = false
+                self.lockIndicatorView.layoutIfNeeded()
+                self.centerContainer.insertArrangedSubview(self.recordingIndicatorView, at: 0)
+                self.centerContainer.insertArrangedSubview(self.slideToCancelView, at: 1)
+            }
```

Comment:
Nit: Since for UIView.animate there are no risks of memory leaks, we could avoid using `weak self` to make the code a bit cleaner and shorter, but no big deal at all 👍

Reply:
Given the duration of the animation there are always risks(the animation may be on flight but the user will swipe back). Additionally I believe it's a good practice to have the weak self in closures. The difference in code will be one line (as `self.` will still be required). If you don't have strong preferences/arguments I would like to keep it otherwise we can make the change in another PR as it will require changes in a few places.

Reply:
Not a big deal at all, but the behaviour you are describing I believe it should not have any significant impact 🤔  In fact, by using `guard let self = self`, it will have the same behaviour you are trying to avoid (AFAIK). If `VoiceRecordingVC` is deallocated after the `guard let self = self`, only after the closure is finished executing is when `VoiceRecordingVC` is completely deallocated. I've actually had issues with this in the past, and it is quite confusing, but `guard let self = self` is creating a strong reference inside that closure execution, so only when the closure finishes executing is when `VoiceRecordingVC` 100% deallocated 

I think it is nice to keep this in mind for the future, but for now, no need to change this since it is not that important and it won't change anything 👍